### PR TITLE
Typescript Jeopardy, Blackjack, and Room Events

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

File: server/chat-plugins/blackjack.ts [hunk 1]
```diff
@@ -38,22 +80,26 @@ class Blackjack extends Rooms.RoomGame {
 		};
 		this.deck = new BlackjackDeck().shuffle();
 
-		this.roomid = this.room.roomid;
+		this.roomID = this.room.roomid as RoomID;
```
Comment (TheJetOU) [inline server/chat-plugins/blackjack.ts]: Is the `RoomID` type cast necessary? I'm almost positive that `Room#roomid`'s type is `RoomID`.

File: server/chat-plugins/blackjack.ts [hunk 2]
```diff
@@ -398,53 +448,57 @@ class Blackjack extends Rooms.RoomGame {
...
 
 		if (player.cards.length < 3) return;
 
 		let turnLine = Chat.html`<br /><strong>${player.name}</strong> hit and received [${player.cards[player.cards.length - 1]}] (${player.points})`;
 		this.turnLog += turnLine;
-		if (player.cards.length > 2) this.display(turnLine, null, player.name);
+		if (player.cards.length > 2) this.display(turnLine, false, player.name);
 
-		if (player === this.dealer) {
+		if (player instanceof BlackjackDealer) {
```
Comment (TheJetOU) [inline server/chat-plugins/blackjack.ts]: `instanceof` breaks in hotpatches. I don't see anything wrong with the original if condition, anyway.

File: server/chat-plugins/blackjack.ts [hunk 3]
```diff
@@ -398,53 +448,57 @@ class Blackjack extends Rooms.RoomGame {
...
 			player.status = 'stand';
-			let turnLine = Chat.html`<br /><strong>${player.name}</strong> has blackjack!`;
+			turnLine = Chat.html`<br /><strong>${player.name}</strong> has blackjack!`;
 			this.turnLog += turnLine;
-			this.display(turnLine, null, player.name);
+			this.display(turnLine, false, player.name);
 			this.clear();
 		}
-		if (user !== 'dealer') this.playerTable[user].cards = player.cards;
+		if (!(player instanceof BlackjackDealer)) this.playerTable[userid].cards = player.cards;
```
Comment (TheJetOU) [inline server/chat-plugins/blackjack.ts]: Ditto.

File: server/chat-plugins/blackjack.ts (hunk 2 above)
Comment (KrisXV) [reply server/chat-plugins/blackjack.ts]: honestly dont remember makign this change

**Comment by Zarel** [comment]
I only skimmed it, but presumably we can fix things if they come up.

File: server/rooms.ts [hunk 4]
```diff
@@ -150,9 +152,11 @@ export abstract class BasicRoom {
 		this.staffRoom = false;
 		this.language = false;
 		this.slowchat = false;
+		this.events = {};
```
Comment (Zarel) [inline server/rooms.ts]: Is this hotpatchable?

Comment (KrisXV) [reply server/rooms.ts]: yep

**Pull request merged**